### PR TITLE
fix for git secret hide -m doesn't work on first use #466

### PR DESCRIPTION
**Pull request by simbo1905**
What does this implement/fix? Explain your changes.
---------------------------------------------------

Ensures that hidden files always have hashes in fsdb. 

Does this close any currently open issues?
------------------------------------------

#466 git secret hide -m doesn't work on first use

Any relevant logs, error output, etc?
-------------------------------------
(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

Any other comments?
-------------------

N/A

### CODE REVIEW

File: tests/test_hide.bats [hunk 1]
```diff
@@ -193,6 +194,37 @@ function teardown {
...
+  cp "${path_mappings}" "${path_mappings}.bak"
+  # run hide again
+  run git secret hide -m
+  # compare
+  [ "$status" -eq 0 ]
+  [[ "${#lines[@]}" -eq 1 ]]
+  
+  # output says 0 of 1 files are hidden because checksum didn't change and we didn't need to hide it again.
+  [ "$output" = "git-secret: done. 0 of 1 files are hidden." ]
+  # no changes should occur to path_mappings files
```
[the remark this answers is not in the document]
Comment (simbo1905) [reply tests/test_hide.bats]: this is a direct copy of the test above it. reading up i don't think it needs square braces as it isn't a test its a call to a utility that might fail. for example when i was googling for 'bats string contains' i came across a recommendation to just pipe command output to grep; if the grep returns none zero the test fails. so running any commands to do work or check the work without square braces seems okay. yet i am very new to bats so I may be totally wrong. as i was copying the existing test above so i would prefer to just keep the existing style and not change this new test then also have to change the tests that it was a direct clone of.

File: tests/test_hide.bats [hunk 2]
```diff
@@ -193,6 +194,37 @@ function teardown {
...
+  # back path mappings
+  cp "${path_mappings}" "${path_mappings}.bak"
+  # run hide again
+  run git secret hide -m
+  # compare
+  [ "$status" -eq 0 ]
+  [[ "${#lines[@]}" -eq 1 ]]
+  
+  # output says 0 of 1 files are hidden because checksum didn't change and we didn't need to hide it again.
+  [ "$output" = "git-secret: done. 0 of 1 files are hidden." ]
```
[the remark this answers is not in the document]
Comment (simbo1905) [reply tests/test_hide.bats]: Awesome! I was looking for how to do a clean 'string contains' and the only hit I got was this on a [bats issue](https://github.com/sstephenson/bats/issues/106) and it didn't have what you suggested. I have posted a comment for other folks.

File: tests/test_hide.bats [hunk 3]
```diff
@@ -50,11 +50,12 @@ function teardown {
...
 
 @test "run 'hide' normally with SECRETS_VERBOSE=1" {
-  SECRETS_VERBOSE=1 run git secret hide
+  SECRETS_VERBOSE=1 run git secret hide 
 
   # Command must execute normally. 
   [ "$status" -eq 0 ]
-  [[ "$output" == "git-secret: done. 1 of 1 files are hidden." ]]
+
+  echo "$output" | grep "git-secret: done. 1 of 1 files are hidden."
```
[the remark this answers is not in the document]
Comment (simbo1905) [reply tests/test_hide.bats]: ditto on the awesome. thanks.

**Comment by simbo1905** [comment]
I have updated the PR. thanks.

File: tests/test_hide.bats (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshrabinowitz) [reply tests/test_hide.bats]: Hm, as I read it `cmp -s` is specifically to compare the contents of the two files (`Output nothing; yield exit status only`) if they can't be different what's the purpose of the command?

However you are correct that other tests seemed to be ignoring the exit code of `cmp` as well so this is not only an issue with the test you added.

I think this is also affected by the SECRETS_VERBOSE setting, which would make `reveal` also show output from the underlying gpg command(s)

File: tests/test_hide.bats [hunk 4]
```diff
@@ -172,7 +172,38 @@ function teardown {
...
+  local path_mappings
+  path_mappings=$(_get_secrets_dir_paths_mapping)
+  run git secret hide
+
+  #echo "$output" | sed "s/^/# '$BATS_TEST_DESCRIPTION' output: /" >&3
+
+  # Command must execute normally:
+  [ "$status" -eq 0 ]
+  # git secret hide -m: uses temp file so cleaning should take place, but we only show tmp file cleanup in VERBOSE mode
+  [[ "${lines[0]}" == *"git-secret: done. 1 of 1 files are hidden."* ]]
```
[the remark this answers is not in the document]
Comment (simbo1905) [reply tests/test_hide.bats]: This is logic within the tests before I fixed the bug. Tests seem to be explicit about when they run verbose or not. I do get the Boy Scout rule about always leaving things in a better state but having to make a lot of changes that didn't exist in the tests before getting a bug passed raises the effort to beyond where it is rewarding to fix bugs. Other than getting the build to pass I am done with this PR. Perhaps a fresh ticket can be raised to make all the tests improved, consistent and defensive?

[the remark this answers is not in the document]
Comment (joshrabinowitz) [reply tests/test_hide.bats]: Wait, there's been a misunderstanding -- for now, it's true, I'm only really concerned with newly added code and the tests that are failing now. I thought that this line was causing the test failures.

File: tests/test_hide.bats (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshrabinowitz) [reply tests/test_hide.bats]: That the exit code  `cmp -s` should be checked is noted in #472

File: tests/test_hide.bats (hunk 4 above)
[the remark this answers is not in the document]
Comment (joshrabinowitz) [reply tests/test_hide.bats]: @simbo1905 just to make sure, you've seen the `osx-with-debug-output` test, which enables SECRETS_VERBOSE, at https://travis-ci.org/sobolevn/git-secret/jobs/525848578 ?

**Comment by joshrabinowitz** [comment]
I'm cool with merging this once it passes tests.

**Comment by simbo1905** [comment]
@joshrabinowitz i am really not sure why an unrelated test fails. Any ideas?

**Comment by joshrabinowitz** [comment]
@simbo1905 I _suspect_ calling the function now called`_fsdb_update_hash` in more cases creates extra output to stdout when TEST_VERBOSE is set and `hide` is used.

**Pull request merged**